### ISSUE TITLE
useEntityRecords to return array rather than object

### ISSUE THREAD

**Issue by mburridge** [WordPress/gutenberg#47222]
## What problem does this address?

[`useEntityRecords`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-core-data/#useentityrecords) currently returns an object, so if you have more than one instance of `useEntityRecords` in a component you need to alias the properties, e.g.: 

```
const { records, isResolving } = useEntityRecords( 'postType', 'post' );
const { records:cats, isResolving:isCatResolving } = useEntityRecords( 'taxonomy', 'category' );
```

## What is your proposed solution?

`useEntityRecords` could instead return an array, similar to what [the native `useState` hook](https://reactjs.org/docs/hooks-state.html) does, so that if there is more than one instance separate names can be assigned to the destructured variables.

```
const [ records, isResolving ] = useEntityRecords( 'postType', 'post' );
const [ cats, isCatResolving ] = useEntityRecords( 'taxonomy', 'category' );
```
This would make for a better developer experience by avoiding the need for aliasing object properties.

This could similarly apply to [`useEntityRecord`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-core-data/#useentityrecord).

cc: @adamziel

**Comment by ryanwelcher** [comment]
Big ➕ from me here. There will be the issue of backward compatibility to address for usages currently out in the community but it feels like the correct move and maybe moving forward any hook that returns generic data like this, should wrap its return values in an array.

Is there a best-practice in place for creating Hooks at the project level?

**Comment by ryanwelcher** [comment]
It looks like now would be a good time to make the changes based on [this search of wpdirectory.net](https://wpdirectory.net/search/01GQ0D6XQVQDC4AQ0R5WPM5SR0) for `useEntityRecords`

tl;dr there 3 plugins that use it

**Comment by Mamaduka** [comment]
The `useEntityRecords` return value object contains four pieces of data, which is the reason for choosing object vs. array.

## Why

When destructuring object order doesn't matter, you can pick and choose which values to extract. But with arrays, order matters; say you want to access 4th element of an array, then you'll have to extract the first three as well.

Here's an example of how you would access the records and status from the hook, object vs. array:

```js
const { records, status } = useEntityRecords( 'postType', 'post' );

// If we decide to change to return value type.
const [ records, , , status ] = useEntityRecords( 'postType', 'post' );
```

Returning an object is a common pattern with similar data-fetching hooks.

TanStack Query: https://tanstack.com/query/v4/docs/react/guides/queries
SWR: https://swr.vercel.app/docs/data-fetching


**Returned interface by useEntityRecords**
```ts
interface EntityRecordsResolution< RecordType > {
	/** The requested entity record */
	records: RecordType[] | null;

	/**
	 * Is the record still being resolved?
	 */
	isResolving: boolean;

	/**
	 * Is the record resolved by now?
	 */
	hasResolved: boolean;

	/** Resolution status */
	status: Status;
}
```

**Comment by adamziel** [comment]
To add to @Mamaduka's excellent response, here's the recommended usage:

```js
const posts = useEntityRecords( 'postType', 'post' );
const cats = useEntityRecords( 'taxonomy', 'category' );
// posts.records
// cats.isResolving
```

**Comment by mburridge** [comment]
Thanks both, that explanation makes perfect sense and the recommended usage is nice and clean and a better DX than aliasing. I'll create an issue to get this properly documented in the [handbook](https://developer.wordpress.org/block-editor/).

**Comment by ryanwelcher** [comment]
Very cool! Thanks for sharing @Mamaduka @adamziel !
